core/templates/dev/head/pages/exploration_editor/translation_tab/TranslationRecordingService.ts

```diff
@@ -0,0 +1,169 @@
+        mp3Worker = new Worker(url);
+        mp3Worker.onmessage = function(e) {
+          switch (e.data.cmd) {
+            case 'end':
+              // async data flow
+              defer.resolve(e.data.buf);
+              break;
+            default :
+              console.warn('Unexpected command and data type from worker.');
+              // console.warn(e);
+          }
+        };
+        mp3Worker.postMessage({cmd: 'init', config: config});
+      }
+    };
+
+    var _postMessage = function(buffer) {
+      mp3Worker.postMessage({cmd: 'encode', buf: buffer});
+    };
+
+    var _stopWorker = function() {
+      if (mp3Worker) {
+        mp3Worker.terminate();
+        console.warn('Ending mp3 worker');
+        mp3Worker = null;
+      }
+    };
+
+    var _initRecorder = function() {
+      // promise required because angular is async with worker
+      AudioContext = $window.AudioContext || $window.webkitAudioContext;
+      if (AudioContext) {
+        defer = $q.defer();
+        isAvailable = true;
+        _initWorker();
+      } else {
+        isAvailable = false;
+      }
+    };
+
+
+
+    var _processMicAudio = function(stream) {
+      definedAudioContext = new AudioContext();
+      // Settings a bufferSize of 0 instructs the browser
+      // to choose the best bufferSize
+      processor = definedAudioContext.createScriptProcessor(0, 1, 1);
+      // process microphone to mp3 encoding
+      processor.onaudioprocess = _onAudioProcess;
+
+      microphone = definedAudioContext.createMediaStreamSource(stream);
+      // connect custom processor to microphone;
+      microphone.connect(processor);
+      // connect to speakers as destination
+      processor.connect(definedAudioContext.destination);
+    };
+
+    // convert directly from mic to mp3
+    var _onAudioProcess = function(event) {
+      var array = event.inputBuffer.getChannelData(0);
+      _postMessage(array);
+    };
+
+    var _startMicrophone = function() {
+      return navigator.mediaDevices.getUserMedia({audio: true, video: false});
+    };
+
+    var _stopRecord = function() {
+      if (microphone && processor) {
+        // disconnect mic and processor and stop processing
+        microphone.disconnect();
+        processor.disconnect();
+        processor.onaudioprocess = null;
+        // issue command to retrieve converted audio
+        mp3Worker.postMessage({cmd: 'finish'});
+        // stop microphone stream
+        microphoneStream.getTracks().forEach(function(track) {
+          track.stop();
+        });
+        isRecording = false;
+      }
+    };
+
+    var _closeRecorder = function() {
+      _stopWorker();
+    };
+
+    return {
+      initRecorder: function() {
+        _initRecorder();
+      },
+      status: function() {
+        return {
+          isAvailable: isAvailable,
+          isRecording: isRecording
+        };
+      },
+      startRecord: function() {
```

Comment:
Maybe startRecording would be better?

Reply:
Ok

Reply:
I don't see what difference it makes but I changed it so it's done.

Reply:
I don't see what difference it makes but I changed it so it's done.